react/getting_started_with_react/keys_in_react.md

```diff
@@ -0,0 +1,286 @@
+
+- Patterns to avoid when using keys
+
+- Can we use index as a key
+
+
+### Why does React need keys?
+
+Lets imagine you take notes while working through The Odin Project.
+
+Maybe you use index cards, or lined paper.
+
+Now if these notes are in a pile, we can think of them as an `array` and each one having a particular `index` in the list.
+
+The paper at the bottom of the pile is the first note you made or `note[0]`, the one directly on top of it being the second note or `note[1]`.
+
+What happens if we want to reorder the list, or remove a note from the list. Maybe we need to find an old note in the middle of the pile.
+
+If `note[4]` was removed, does that mean `note[5]` just became `note[4]`? does every note after `note[5]` have to be re indexed?
+
+If we reorder or *sort* the notes, we might end up in a situation where our previous `note[14]` might become our new `note[3]`.
+
+It gets a bit confusing to manage and keep track of which papers are still in order.
+
+One solution might be to add a unique summary or `key` of some kind to each page.
+
+This way no matter where that specific `note` moves to in the pile, we know which note it is.
+
+Similarly `keys` help React to keep track of the items in a list. `keys` allow React to know when to render, what to render, when to re render, and when not to re render.
+
+
+### What does it matter if the keys match or do not match?
+
+If an item in an array changes, keys help React decide which dom elements have changed.
+
+If a single item in a list where to change, one of two things *should* happen:
+Either
+We completely re render the entire list
+Or
+Hunt down that one specific item that was changed and only re render that item.
+
+Assuming we want to hunt down that one specific item that was changed and NOT re render the entire list. We need something to *track* that specific item.
+
+We can *track* or *hunt* down a specific item by using a `key`.
+
+If we were to give each item in the list an `identifier` and each item in the dom the corresponding `identifier` it becomes really easy to keep track of that element in both the logic and the dom.
+
+When an item is changed, React uses that item's `key` to identify which item has changed, and which dom element to update.
+
+If the `key` does NOT match the item being changed, react will leave that list item alone.
+If React DOES find a match, it will update that item in the dom accordingly.
+
+As long as `keys` remain consistent and unique, React can handle the dom effectively and efficiently.
+
+### What are keys?
+
+`keys` are `strings`
+
+`keys` connect the DOM item with the item in data.
+
+`keys` help us keep track of our state (logic) and our dom.
+
+`keys` tell React when the dom should be rendered or re rendered.
+
+When list items are moved, added or deleted, React uses `keys` to see which items in the dom need to be moved, added, or deleted.
+
+Without `keys`, React would not be able to handle the dom as elegantly as it does.
```

Comment:
The "What are keys?" section seems redundant. I feel almost all the content written here gets covered elsewhere throughout the lesson.